core/src/main/java/com/scalar/db/transaction/common/AbstractDistributedTransactionManager.java

```diff
@@ -177,8 +184,119 @@ public void rollback() throws RollbackException {
       }
     }
 
+    @Override
+    public DistributedTransaction getOriginalTransaction() {
+      if (transaction instanceof WrappedDistributedTransaction) {
+        return ((WrappedDistributedTransaction) transaction).getOriginalTransaction();
+      }
+      return transaction;
+    }
+  }
+
+  @VisibleForTesting
+  static class TransactionStateManagement extends AbstractDistributedTransaction
+      implements WrappedDistributedTransaction {
+
+    private enum Status {
+      ACTIVE,
+      COMMITTED,
+      COMMIT_FAILED,
+      ROLLED_BACK
+    }
+
+    private final DistributedTransaction transaction;
+    private Status status;
+
     @VisibleForTesting
-    public DistributedTransaction getActualTransaction() {
+    TransactionStateManagement(DistributedTransaction transaction) {
+      this.transaction = transaction;
+      status = Status.ACTIVE;
+    }
+
+    @Override
+    public String getId() {
+      return transaction.getId();
+    }
+
+    @Override
+    public Optional<Result> get(Get get) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      return transaction.get(get);
+    }
+
+    @Override
+    public List<Result> scan(Scan scan) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      return transaction.scan(scan);
+    }
+
+    @Override
+    public void put(Put put) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      transaction.put(put);
+    }
+
+    @Override
+    public void put(List<Put> puts) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      transaction.put(puts);
+    }
+
+    @Override
+    public void delete(Delete delete) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      transaction.delete(delete);
+    }
+
+    @Override
+    public void delete(List<Delete> deletes) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      transaction.delete(deletes);
+    }
+
+    @Override
+    public void mutate(List<? extends Mutation> mutations) throws CrudException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      transaction.mutate(mutations);
+    }
+
+    @Override
+    public void commit() throws CommitException, UnknownTransactionStatusException {
+      checkStatus("The transaction is not active", Status.ACTIVE);
+      try {
+        transaction.commit();
+        status = Status.COMMITTED;
+      } catch (Exception e) {
+        status = Status.COMMIT_FAILED;
+        throw e;
+      }
+    }
+
+    @Override
+    public void rollback() throws RollbackException {
+      if (status == Status.COMMITTED || status == Status.ROLLED_BACK) {
```

Comment:
Allow to call this `rollback()` method only when the status is `ACTIVE` or `COMMIT_FAILED`.